### PR TITLE
Consensus: fix panic in batch verifier

### PR DESCRIPTION
**Pull request by Boog900**
This issue was already documented in the batch verifier but I didn't realize it was going to be so common, me and @SyntheticBird45 have both hit this panic.

### CODE REVIEW

File: consensus/rules/src/batch_verifier.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+use multiexp::BatchVerifier as InternalBatchVerifier;
+
+/// This trait represents a batch verifier.
+pub trait BatchVerifier {
```
Comment (hinto-janai) [inline consensus/rules/src/batch_verifier.rs:4]: What is a batch verifier, why is it needed, what does it do, etc.

File: consensus/rules/src/batch_verifier.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    /// Queue a statement for batch verification.
+    ///
+    /// # Panics
+    /// This function may panic if stmt contains calls to rayon par_iters.
```
Comment (hinto-janai) [inline consensus/rules/src/batch_verifier.rs:8]:
```suggestion
    /// This function may panic if `stmt` contains calls to `rayon`'s parallel iterators, e.g. `par_iter()`..
```

File: consensus/rules/src/transactions/ring_ct.rs [hunk 3]
```diff
@@ -91,7 +90,7 @@ fn simple_type_balances(rct_sig: &RctSignatures) -> Result<(), RingCTError> {
 /// <https://monero-book.cuprate.org/consensus_rules/ring_ct/bulletproofs+.html>
 fn check_output_range_proofs(
     rct_sig: &RctSignatures,
-    verifier: &mut BatchVerifier<(), dalek_ff_group::EdwardsPoint>,
+    mut verifier: impl BatchVerifier,
```
Comment (hinto-janai) [inline consensus/rules/src/transactions/ring_ct.rs:93]: Can/should this be a reference?
```rust
verifier: &mut impl BatchVerifier,
```

File: consensus/src/batch_verifier.rs [hunk 4]
```diff
@@ -41,3 +26,17 @@ impl MultiThreadedBatchVerifier {
             .is_none()
     }
 }
+
+impl<'a> cuprate_consensus_rules::batch_verifier::BatchVerifier for &'a MultiThreadedBatchVerifier {
```
Comment (hinto-janai) [inline consensus/src/batch_verifier.rs:30]:
```suggestion
impl cuprate_consensus_rules::batch_verifier::BatchVerifier for &'_ MultiThreadedBatchVerifier {
```

File: consensus/rules/src/transactions/ring_ct.rs (hunk 3 above)
Comment (Boog900) [reply consensus/rules/src/transactions/ring_ct.rs:93]: I made the trait impl for the single threaded verifier `&mut`

File: consensus/rules/src/batch_verifier.rs (hunk 1 above)
Comment (Boog900) [reply consensus/rules/src/batch_verifier.rs:4]: added some docs

**Review by hinto-janai — APPROVED** [review]

File: consensus/rules/src/batch_verifier.rs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+use multiexp::BatchVerifier as InternalBatchVerifier;
+
+/// This trait represents a batch verifier.
+///
+/// A batch verifier is used to speed up verification by verifying multiple transactions together.
+///
+/// Not all proofs can be batched and at it's core it's intended to verify a series of statements are
```
Comment (hinto-janai) [inline consensus/rules/src/batch_verifier.rs:7]: I... can't... help it ahhhhhhhhhhh

it's = it is
its = possession

I'm sorry in advance.

```suggestion
/// Not all proofs can be batched and at its core it's intended to verify a series of statements are
```

**Pull request merged**